### PR TITLE
support for lambda functions

### PR DESCRIPTION
**Pull request by HliasMpGH**
*testing
*documentation

### CODE REVIEW

**Comment by HliasMpGH** [comment]
This PR issues changes such as:

CodeBlock & MethodSpec:
-addLambda() overloads used for constructing different types of lambda functions.
-testing for different cases.

A plugin addition to the build, used for producing javadoc from source.
Some reformatting of imports and javadoc.
A bug fix on TypeSpec where an overload of addPermits(), created an infinite recursion.

File: src/main/java/com/squareup/javapoet/TypeSpec.java [hunk 1]
```diff
@@ -609,7 +609,7 @@ public Builder addSuperinterface(Type superinterface, boolean avoidNestedTypeNam
     public Builder addPermits(Iterable<? extends TypeName> permits) {
       checkArgument(permits != null, "permits == null");
       for (TypeName permit : permits) {
-        addPermits(permits);
+        addPermits(permit);
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/TypeSpec.java:612]: nit: I'd put this in a separate commit as it's not related to this commit

**Comment by Randgalt** [comment]
Thanks - I'll review as I have time. I'm not very familiar with the JPoet code base but I'll do my best. Maybe we can get some others to review from the community.

File: src/main/java/com/squareup/javapoet/TypeSpec.java (hunk 1 above)
Comment (HliasMpGH) [reply src/main/java/com/squareup/javapoet/TypeSpec.java:612]: Must have missed it while squashing my commits, sorry for the inconvenience. Thanks!

File: pom.xml [hunk 2]
```diff
@@ -135,6 +135,16 @@
     <build>
         <pluginManagement>
             <plugins>
+                <plugin>
```
Comment (Randgalt) [inline pom.xml:138]: Why is this needed? If just personal preference please put in a separate commit with its own comment.

File: src/main/java/com/squareup/javapoet/TypeSpec.java (hunk 1 above)
Comment (Randgalt) [reply src/main/java/com/squareup/javapoet/TypeSpec.java:612]: Have you pushed latest? I'm still seeing this in the 1 commit.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 3]
```diff
@@ -415,6 +417,81 @@ public Builder add(CodeBlock codeBlock) {
...
+     * Should be used with {@link #add(String, Object...) add},
+     * to provide specific behaviour such as
+     * methodcall((int x, int y) -> {return x + y;}, 5).
+     * @param parameters the input parameters of the function.
+     * @param expressionFromat the format that should be used
+     * for the expression.
+     * @param args the values that should be placed in the holders
+     * of the format.
+     */
+    public Builder addLambda(List<ParameterSpec> parameters, String expressionFromat, Object... args) {
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:462]: Spelling: `expressionFromat` should be `expressionFormat` (multiple places)

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 4]
```diff
@@ -415,6 +417,81 @@ public Builder add(CodeBlock codeBlock) {
...
+      // check that the input types are valid
+      for (ParameterSpec parameter : parameters) {
+        checkArgument(!parameter.type.equals(TypeName.VOID),
+          "lambda input parameters cannot be of void type!");
+      }
+
+      // the inputs of the lambda (left side)
+      String inputSide = String.join(
+        ", ", parameters.stream()
+        .map(p -> p.toString())
```
[the remark this answers is not in the document]
Comment (Randgalt) [reply src/main/java/com/squareup/javapoet/CodeBlock.java:439]: Whole thing can be:

```suggestion
      String inputSide = parameters.stream()
      .map(ParameterSpec::toString)
      .collect(Collectors.joining(", "));
```

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 5]
```diff
@@ -415,6 +417,81 @@ public Builder add(CodeBlock codeBlock) {
...
+     * to provide specific behaviour such as
+     * methodcall((int x, int y) -> {return x + y;}, 5).
+     * @param parameters the input parameters of the function.
+     * @param body the body of the function.
+     */
+    public Builder addLambda(List<ParameterSpec> parameters, CodeBlock body) {
+      // check that the input types are valid
+      for (ParameterSpec parameter : parameters) {
+        checkArgument(!parameter.type.equals(TypeName.VOID),
+          "lambda input parameters cannot be of void type!");
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:433]: nit: the rest of JavaPoet has message that does not use `!`

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 6]
```diff
@@ -415,6 +417,81 @@ public Builder add(CodeBlock codeBlock) {
...
+    /**
+     * Structures a lambda expression containing a body,
+     * and not only an expression.<br>
+     * Should be used with {@link #add(String, Object...) add},
+     * to provide specific behaviour such as
+     * methodcall((int x, int y) -> {return x + y;}, 5).
+     * @param parameters the input parameters of the function.
+     * @param body the body of the function.
+     */
+    public Builder addLambda(List<ParameterSpec> parameters, CodeBlock body) {
```
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:429]: To be consistent with the rest of JavaPoet these should be `Iterable<ParameterSpec>`. You can convert to a stream via:

```
Stream<ParameterSpec> parameterSpecStream = StreamSupport.stream(parameters.spliterator(), false);
```

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 3 above)
Comment (Randgalt) [inline src/main/java/com/squareup/javapoet/CodeBlock.java:462]: Same as above, make `Iterable` to match rest of the library

**Comment by Randgalt** [comment]
I have a general comment about this. I think we should consider a more complete solution. Right now, this modifies CodeBlock and generates a lambda like this:

```
(double y) -> {return x + y;}
```

This is good but not ideal. If I were to write the above lambda, I'd write:

```
y -> x + y;
```

Instead of modifying CodeBlock with addLambda we probably need a `LambdaSpec` similar to `MethodSpec` that can analyze the parameters, return type, etc. and build a lambda more like what developers expect.

wdyt?

**Comment by HliasMpGH** [comment]
> I have a general comment about this. I think we should consider a more complete solution. Right now, this modifies CodeBlock and generates a lambda like this:
> 
> ```
> (double y) -> {return x + y;}
> ```
> 
> This is good but not ideal. If I were to write the above lambda, I'd write:
> 
> ```
> y -> x + y;
> ```
> 
> Instead of modifying CodeBlock with addLambda we probably need a `LambdaSpec` similar to `MethodSpec` that can analyze the parameters, return type, etc. and build a lambda more like what developers expect.
> 
> wdyt?


You are right. My initial thought was to follow the LambdaSpec class idea, and that would be required if we want to allow code like
```
x -> x + 3;
```

The reason i avoided the ideas was because it appeared to me that maybe it wouldnt fit with the rest of the Spec's of the project and was an an overkill to have a separate class for that, thats why i went through with the current implementation. That doesnt mean though that LambdaSpec cant be a thing, and since you propose it as well, i can make the changes needed to support it. My idea of the final code will be something like this:
```
ParameterSpec consumerLambda = ParameterSpec.builder(
	ParameterizedTypeName.get(
				ClassName.get(Consumer.class),
				TypeName.get(Integer.class)
			),"consumer"
		).build();

ParameterSpec p = ParameterSpec.builder(TypeName.INT, "x").build();

CodeBlock body = CodeBlock.builder().add("x + 5").build();

LambdaSpec lambda = LambdaSpec.builder(consumerLambda).addInput(p).addBody(body).build();
```
that will output something like this:
```
Consumer<Integer> consumer = x -> x + 5;
```
or 
```
LambdaSpec lambda = LambdaSpec.builder().addInput(p).addBody(body).showTypes().build();
```
to produce:
```
(int x) -> x + 5;
```
for more general use (for example if we need to pass it as input to a method).

We would then have methods such as addLambda(LambdaSpec) in CodeBlock, that would be used for the final construction.

Does that make sense?

File: pom.xml (hunk 2 above)
Comment (HliasMpGH) [reply pom.xml:138]: I will make sure that every change apart from the lambdas will be in a separate commit

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (HliasMpGH) [reply src/main/java/com/squareup/javapoet/CodeBlock.java:439]: Ah, i see. Thank you for pointing it out!

**Comment by Randgalt** [comment]
@HliasMpGH I don't think a `LambdaSpec` builder is necessary if we can get the output to be ideal with only a `CodeBlock`. wdyt?

**Comment by HliasMpGH** [comment]
> Sorry, no, I meant keep what you have but make it be able to build an ideal statement. I'm not sure if it's possible.

Ohh i see. Okay, i will do my best to try and add that ability.

**Comment by HliasMpGH** [comment]
Closing to reduce pollution and clear up space. The implemented result can be seen in #8
